### ISSUE TITLE
Generic Host

### ISSUE THREAD

**Issue by davidfowl** [aspnet/Hosting#1163]
## Generic Host

As part of ASP.NET Core, we introduced a set of cross cutting concerns that are fundamental to building many applications.
These include things like configuration, dependency injection, and logging. The `WebHostBuilder` is the glue that brings those common cross cutting
concerns together but it also couples the building of the HTTP pipeline into the same component. The goal of the generic host is to de-couple
the HTTP pipeline building from the Host API to enable other scenarios like messaging, background tasks, and supporting other non HTTP workloads that would benefit
from the same cross-cutting libraries that HTTP benefits from.

### Introducing Microsoft.Extensions.Hosting

Microsoft.Extensions.Hosting is the generic hosting library that will offer a similar API to `WebHostBuilder` but won't be specific to HTTP.
The eventual goal is to completely replace the `WebHostBuilder` API with this new one. The generic host will offer a few building blocks
for application models:

#### IHostedService

```C#
public interface IHostedService
{
    Task StartAsync(CancellationToken cancellationToken);

    Task StopAsync(CancellationToken cancellationToken);
}
```

This is the entry point to code execution. Many of these may be registered in as services. They will be resolved and executed **in order of registration**. StartAsync will be called when the host starts, and StopAsync will be called in **reverse registration order** when the host shuts down gracefully. IHostedServices are registered as part of container registration and resolved by the host as an `IEnumerable<IHostedService>`.

#### IHostBuilder

```C#
public interface IHostBuilder 
{
    IHost Build();
    
    IHostBuilder ConfigureAppConfiguration(Action<HostBuilderContext, IConfigurationBuilder> configureDelegate);

    IHostBuilder ConfigureHostConfiguration(Action<IConfigurationBuilder> configureDelegate);
    
    IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureServices);

    IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderContext, TContainerBuilder> configureContainer);

    IHostBuilder UseServiceProviderFactory<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory);
    
    IDictionary<object, object> Properties { get; }
}
```

This is the main component that libraries and applications will extend to light up functionality.

#### IHost

```C#
public interface IHost : IDisposable
{
   IServiceProvider Services { get; }
   Task StartAsync(CancellationToken cancellationToken = default(CancellationToken));
   Task StopAsync(CancellationToken cancellationToken = default(CancellationToken));
}
```

The `IHost` implementation is responsible for starting and stopping all the `IHostedServices` that have been registered in DI.

There will also be a set of default services offered by the hosting environment:

#### IHostingEnvironment

```C#
public interface IHostingEnvironment
{
    string EnvironmentName { get; set; }
    
    string ApplicationName { get; set; }

    string ContentRootPath { get; set; }

    IFileProvider ContentRootFileProvider { get; set; }
    
    IConfiguration Configuration { get; }
}
```

The `IHostingEnvironment` provides information about the hosting environment an application is running in.

#### IApplicationLifetime

```C#
public interface IApplicationLifetime
{
    CancellationToken ApplicationStarted { get; }

    CancellationToken ApplicationStopping { get; }

    CancellationToken ApplicationStopped { get; }

    void StopApplication();
}
```

The `IApplicationLifetime` is a service that allows the application to handle host lifetime events and also request graceful shutdown of the host.

#### ILoggerFactory

https://github.com/aspnet/Logging

#### IConfiguration

https://github.com/aspnet/Configuration

#### IOptions\<T\>

https://github.com/aspnet/Options

### Extensibliltiy

Host extensibility is done via extension methods on the `IHostBuilder`. The builder provides the ability to manipulate the core primitives and the 
expectation is that extension methods can use this to build higher level functionality:

```C#
// UseRabbitMq is an extension method that registers a rabbit MQ 
// IHostedService and handles incoming messages.
var host = new HostBuilder()
            .UseRabbitMq<MyMessageHandler>() 
            .Build();
            
await host.StartAsync();
```

### DependencyInjection

In order to support plugging in other containers, the host can accept an [IServiceProviderFactory](https://github.com/aspnet/DependencyInjection/blob/06e2de235ce5b27b425e823d9dcbd045811ba48e/src/DI.Abstractions/IServiceProviderFactory.cs). This itself will *not* be part of the DI container registration, but will be a host intrinsic used to create the concrete DI container. 

Custom container configuration can be done via the ConfigureContainer method. This gives you a strongly typed experience for configuring your container on top of the underlying API.

### Host Lifetime

In order to accomodate systems that control the lifetime of the host, we introduce an new interface `Lifetime` to delegate lifetime management to another system for e.g. windows service host or console host. 

```C#
public interface IHostLifetime
{
    void OnStarted(Action<object> callback, object state);
    void OnStopping(Action<object> callback, object state);
    Task StopAsync();
}
```

When `IHost.StartAsync` is called, it will register with the `IHostLifetime` to notify the caller when the application started or stopped via `OnStarted` and `OnStopping`. When `IHost.StopAsync` is called, the host will notify the `IHostLifetime` that it has stopped.

Here's an example implementation of the `ConsoleHostLifetime` that triggers shutdown on Control + C:

```C#
public class ConsoleLifetime : IHostLifetime
{
    public void OnStarted(Action<object> callback, object state)
    {
        // There's no event to wait on started
        callback(state);
    }

    public void OnStopping(Action<object> callback, object state)
    {
        Console.CancelKeyPress += (sender, e) =>
        {
            e.Cancel = true;
            callback(state);
        };
    }

    public Task StopAsync()
    {
        // There's nothing to do here
        return Task.CompletedTask;
    }
}
```

### Startup

*A note about `Startup` classes* - While the concept of a Startup class is extremely useful, it's also very problematic as it requires
2 dependency injection containers to be built during the startup cycle. This causes issues where singletons aren't the same when activating the
Startup class and when activating an `IHostedService`. This is because Startup.ConfigureServices lets the user add more services but also lets them
get at hosting services in the Startup.ctor. So far, we've chosen to abandon startup as part of the generic host work but it would be possible to add back
at any point in the future if we figure out a cleaner way to do it.

### Open Questions

- We only expose a Start and Stop event but other hosts might have more advanced events they want to surface. Windows store has suspend/resume, windows services can be paused and continued as well. Do we need to plan for expanding the life cycle or is that out of scope?
- How do we gradually migrate people over to this new pattern from `WebHostBuilder`?
- Do we need to support ordering of IHostedServices? What if people want to write components that run after servers have started?
- Do we need a first class server component that runs after hosted services (related to the previous bullet)?
- The `IServiceProviderFactory<TContainer>` interface being generic might cause problems (no way to flow the generic through the system). We might need a non-generic version /cc @pakrym

**Comment by nwoolls** [comment]
Are `IHostBuilder.ConfigureAppConfiguration()` and `IHostBuilder.ConfigureAppConfiguration()` meant to return `IWebHostBuilder`?

**Comment by davidfowl** [comment]
Fixed! Thanks @nwoolls

**Comment by ReubenBond** [comment]
Thanks for opening the discussion, @davidfowl. @jdom pointed me here.

Glad that `Startup` class is gone. If a user wants to bring their own container, how do they do this using this new abstraction? I'm currently prototyping `SiloBuilder` for Orleans 2.0. This is what I have so far:
```C#
public interface ISiloBuilder
{
    ISilo Build();

    ISiloBuilder ConfigureServices(Action<IServiceCollection> configureServices);

    ISiloBuilder ConfigureServiceProvider(Func<IServiceCollection, IServiceProvider> configureServiceProvider);
}
```

`ConfigureServiceProvider` can be used to modify the services collection one last time and build the container.

Can `IHost` signal that it has terminated on its own accord? Self-termination is fairly common in some kinds of distributed systems where a node might decide to terminate for safety/correctness purposes. The typical pattern is that the hosting environment would restart it. This is my current equivalent to `IHost`:
```C#
public interface ISilo
{
    Task StartAsync(CancellationToken cancellationToken = default(CancellationToken));

    // If the cancellation token is already canceled, the silo will terminate ungracefully.
    // Otherwise it will perform a graceful shutdown, transfering its responsibilities to
    // other silos.
    Task StopAsync(CancellationToken cancellationToken = default(CancellationToken));

    IServiceProvider Services { get; }

    // The silo can kill itself and signal termination via this task.
    Task Stopped { get; }
}
```

In the event that an `IHostedService` terminates out-of-band, I expect that all others are stopped. Does that sound right?

If an `IHostedService` wants to receive notification that it should start graceful shutdown, is registering with `IApplicationLifetime` the mechanism for that?

Is `IHostBuilder.Properties == IHostingEnvironment.Properties`? If not, what's the difference?

**Comment by davidfowl** [comment]
> ConfigureServiceProvider can be used to modify the services collection one last time and build the container.

This looks like our IServiceProviderFactory concept. https://github.com/aspnet/DependencyInjection/blob/dev/src/DI.Abstractions/IServiceProviderFactory.cs. Thanks for the reminder, this is important for replacing the DI container.


> Can IHost signal that it has terminated on its own accord? Self-termination is fairly common in some kinds of distributed systems where a node might decide to terminate for safety/correctness purposes. The typical pattern is that the hosting environment would restart it. This is my current equivalent to IHost:

I believe the equivalent is `IApplicationLifetime`.

> In the event that an IHostedService terminates out-of-band, I expect that all others are stopped. Does that sound right?

Nope. The IHostedService would need to shutdown the host to terminate the other IHostedServices via IApplicationLifetime.StopApplication.

> If an IHostedService wants to receive notification that it should start graceful shutdown, is registering with IApplicationLifetime the mechanism for that?

No need. StopAsync will be called if the application shuts down in any way.

> Is IHostBuilder.Properties == IHostingEnvironment.Properties? If not, what's the difference?

IHostBuilder.Properties just allows you to round trip state during the building of the host. One reason might by idempotency. You wrote an extension methods that performed some modificationto the host and you don't want it to happen multiple times with several calls. You can use the Properties to store state and check it to see if that modification happened.

IHostingEnvironment.Properties are just a grab bag of things the host wants to flow to the application. Sorta like "server variables" or "environment variables". Of course you could just use a first class service for that.

**Comment by jdom** [comment]
This looks great, I think we can align with Orleans too to use this.
One thing that isn't very clear here is what is the replacement for the `StartupType.Configure` method, which executes after the service provider was created? Or is the plan that users will be required to register some built-in (via extension methods) or custom `IHostedService` that can access services from the container as soon as the host starts and then finish the service configuration part?

There was no metion of `Microsoft.Extension.Options`. I assume it's because ASP.NET will reference it, but not the hosting abstraction?

**Comment by davidfowl** [comment]
> One thing that isn't very clear here is what is the replacement for the StartupType.Configure method, which executes after the service provider was created? Or is the plan that users will be required to register some built-in (via extension methods) or custom IHostedService that can access services from the container as soon as the host starts and then finish the service configuration part?

IHostedService is the entry point. As an example, if we had this before `WebHostBuilder`, there would be a WebServiceHostedService that on Start, calling to Startup and then starts the server.

> There was no metion of Microsoft.Extensions.Options. I assume it's because ASP.NET will reference it, but not the hosting abstraction?

That's a good point. We had a discussion on what the bare minimum set of required things needed to being in the core hosting layer. Options came up but we decided that it wasn't strictly required. That of course isn't set in stone. I think it would be valuable to add it as we added logging as well (which you can make the same argument, isn't required).

**Comment by jason-bragg** [comment]
@davidfowl Just for consideration in regards to the lifetime management and how it interacts with hosted services, we (Orleans team) prototyped this some ourselves as we needed better control over the per server startup/shutdown process.  This is especially relevant to us because our hosts (silos) are part of a cluster that is intended to function as a whole for distributed applications.

The pattern we've come up with (and are not 100% committed to yet) consists of an observable lifetime (which we called lifecycle) and a lifecycle observer, which hosted services (or other components) could act as.  These interfaces are as follows:

    public interface ILifecycleObservable<in TStage>
    {
        IDisposable Subscribe(TStage stage, ILifecycleObserver observer);
    }

    public interface ILifecycleObserver
    {
        Task OnStart(CancellationToken ct);
        Task OnStop(CancellationToken ct);
    }

The stages of the lifecycle are started in order, and stopped in reverse order, providing a staged startup process that allows dependent systems to be started and coordinated (even across a cluster) before the host is considered 'active' and ready for external traffic.

Our ILifecycleObservable is notably similar to the IHostLifetime, so I strongly suspect we'll be able to use our lifecycle pattern in a host built using what's been described here, but, as a general pattern, the lifecycle we're tinkering with is not limited to host lifetime, we're also using it in our grain lifecycle management (and likely any other complex start/stop sequences).

The PR where we introduced this can be seen in https://github.com/dotnet/orleans/pull/3270 should you want to take a deeper look.

I'm not suggesting you change the lifetime pattern suggested in this thread, as it suits us fine, only bringing this similar pattern to your attention for consideration as you refine the host design.

**Comment by pksorensen** [comment]
For my non http service fabric applications my startup looks like this

```
        private static void Main()
        {
            try
            {     
                var config = new ConfigurationBuilder()
                    .SetBasePath(Directory.GetCurrentDirectory())
                    .AddEnvironmentVariables(prefix: "ASPNETCORE_");

                

                using (var container = new FabricContainer())
                {
                    container.AddOptions();
                    container.UseConfiguration(config);
                    container.RegisterType<IHostingEnvironment>(new ContainerControlledLifetimeManager(),new InjectionFactory(c=>
                    {
                        var _options = new WebHostOptions(c.Resolve<IConfiguration>());
                        var contentRootPath = ResolveContentRootPath(_options.ContentRootPath, AppContext.BaseDirectory);
                        var hostingEnvironment = new HostingEnvironment();
                        hostingEnvironment.Initialize("Elastic", contentRootPath, _options);
                        return hostingEnvironment;
                    }));
                    container.RegisterType(typeof(ILogger<>), typeof(Logger<>),new ContainerControlledLifetimeManager());
                    container.ConfigureSerilogging(logConfiguration =>
                             logConfiguration.MinimumLevel.Debug()
                             .Enrich.FromLogContext()
                             .WriteTo.LiterateConsole(outputTemplate: "[{Timestamp:HH:mm:ss} {Level}] {SourceContext}{NewLine}{Message}{NewLine}{Exception}{NewLine}")
                             .WriteTo.ApplicationInsightsTraces("10e77ea7-1d38-40f7-901c-ef3c2e7d48ef", Serilog.Events.LogEventLevel.Information));

                  
                    container.WithStatelessService<ElasticService>("ElasticServiceType");

                    // Prevents this host process from terminating so services keep running.
                    Thread.Sleep(Timeout.Infinite);

                }
                 
               
            }
            catch (Exception e)
            {
                ServiceEventSource.Current.ServiceHostInitializationFailed(e.ToString());
                throw;
            }
        }
```

So to get rid of the self plumbing for the basic stuff, that a generic host would offer, would be nice.

as for 

> A note about Startup classes - While the concept of a Startup class is extremely useful, it's also very problematic as it requires
2 dependency injection containers to be built during the startup cycle. This causes issues where singletons aren't the same when activating the
Startup class and when activating an IHostedService

I think unity dependency injection has a clean way of doing this. Instead of having both a service collection and a service resolver, its all the same. Meaning as soon a type has been registered, it can be resolved right away without building a new serviceProvider. This could solve the stuff about building and registering in startup.

**Comment by davidfowl** [comment]
> So to get rid of the self plumbing for the basic stuff, that a generic host would offer, would be nice.

Nice!

> I think unity dependency injection has a clean way of doing this. Instead of having both a service collection and a service resolver, its all the same. Meaning as soon a type has been registered, it can be resolved right away without building a new serviceProvider. This could solve the stuff about building and registering in startup.

Yea, that sounds like a way to solve it but we're not changing any of the dependency injection system as part of this. I'm also not a fan of having a mutable service provider. That can lead to inefficiencies and bugs.

**Comment by andrebires** [comment]
I work in a messaging platform that implements similar concepts for generic hosting. It currently runs both as Windows services and console applications. I will share the details of how it works and maybe can serve as inspiration in some way. This structure has been used for some years in a very large codebase, working very well for us until now.

First, we have a very basic primitive caller ```IWorker``` that defines a component that execute a background task that can be eventually stopped. The idea is a rip off from TPL Dataflow (which we rely deeply), where all blocks have a ```void Complete()``` method and a ```Task Completion``` property to allow awaiting the job to end. The job in this case is *hot* and starts right after the instantiation of the class. It may be a one-shot job that ends by itself after some time (like running database migrations) or it can be a loop job that ends only if stopped (like listening for connections). The job should considered finished only when the ```Execution``` task is completed and it may hold exceptions in case of any errors during the job.

```csharp
interface IWorker
{
    void Stop();

    Task Execution { get; }
}
```

The ```Stop``` method is synchronous because it only sends a signal to the task to stop it (i.e. cancel a CancellationTokenSource). If there's any async task to be executed after the work is stopped, it should be executed in the task hold by the ```Execution``` property.

Then we have a ```IService``` interface that extends ```IWorker``` and allows the *cold* start of the job, adding the ```StartAsync``` method.

```csharp
interface IService : IWorker
{
    Task StartAsync(CancellationToken cancellationToken);
}
```

The ```StartAsync``` method is asynchronous because we may need to execute some preparation tasks for the job that may fail to execute and compromise the initialization.

We have another primitive called ```IServiceContainer``` that holds and orchestrate the execution of multiple services. It is a service by itself and allows adding services in specific *tiers*.

```csharp
public interface IServiceContainer : IService
{        
    IServiceContainer Add(IService service, int tier = 0);
} 
```

When you start the container, it start all contained services in the order defined by the tier value, from the lower value to the higher. Services in the same tier are started in the same time. The ```Execution``` of the container holds the execution of all child services. When the container is stopped, the higher tiers are stopped first, going to the lower tiers after that. 

Finally, we the the ```IServiceActivator``` interface that defines a service for bootstrapping services. 

```csharp
public interface IServiceActivator : IService
{
    IDictionary<Type, IServiceProvider> ServiceProviders { get; }
    
    ICollection<string> ActivationGroups { get; }        
}
```
The ```ServiceProviders``` property holds the dependency injection containers for each activated service and the ```ActivationGroups```, which acts like a filter for which services should be activated accordingly to the host configuration. Our activator implementation searchs for ```IService``` implementations in the host (Console / Windows Service) referenced assemblies decorated with the ```ActivateAttribute```, using an ```IServiceContainer``` for holding the created instances.

```csharp
 [AttributeUsage(AttributeTargets.Class)]
public sealed class ActivateAttribute : Attribute
{    
    public Type RegistrableType { get; set; }

    public int Tier { get; set; }

    public string ActivationGroup { get; set; }
}
```

Each service is instantiated by its own container, which is created by the activator and populated by the ```IRegistrable``` instance defined in the ```RegistrableType``` property.

```csharp
public interface IRegistrable
{
    void RegisterTo(Container container);
}
```

The ```Container``` class is just an ```System.IServiceProvider``` which also allows the registration of types. The activator instantiates a new container, passes it to the ```RegisterTo``` method and try resolves an ```IService``` instance after that.

We also have a lot of helper classes for this infrastructure, like a ```ServiceBase``` class which implements common patterns for services (like synchronized initialization - avoiding problems when there are multiple start/stop calls) and services for monitoring the health of services (restarting automatically failed services), but the basics are above.

**Comment by tuespetre** [comment]
> ConsoleHostLifetime that triggers shutdown on `Control + C`

This is what I came for and you did not disappoint, @davidfowl. 😂

**Comment by davidfowl** [comment]
@andrebires very cool!

**Comment by andrebires** [comment]
@davidfowl Yes, the service activator creates a DI container for each activated service and expose they throught the ```ServiceProviders``` property . Currently, this is exposed only for testing reasons - sometimes we need to retrieve an instance created in a child container for asserting some conditions - but we could have an ```IServiceActivator``` implementation that, instead of automatically discovering the services to be activated, it receives they throught this property, separating the service discovery of the activation. The activator initialize every service by resolving its type in its own container.

Also, our activator implementation receives in the constructor a root DI container and every created child container has a reference to it, creating an DI hierarchy. If the service requires a type that is not available in its own container, it can be resolved by the root container. This is useful in some cases that we need to share some types between services. For instance, we have a connection listener service that creates "in process" transport connections to our platform, so the client connections produced by this listener should be consumed by other service.

**Comment by jthelin** [comment]
It is important for this spec to be clear about what ordering guarantees [or not] there are related to `StartAsync` / `StopAsync` call sequences. Specifically:

1. Do calls to `IHostedService.StartAsync` happen _sequentially_, or in parallel? I am assuming sequential is the "default" [and much easier to reason about!], but is there also some way to have _some_ services initialize in parallel with each other? [to expedite server startup times]
2. Do calls to `IHostedService.StopAsync` occur in *reverse* order to the `StartAsync` calls? Every system I have ever written what uses similar server-hosting functionality had some notion of "layering" which should be built up "bottom-to-top" and torn down "top-to-bottom".
3. What is the mechanism to define & ensure a specific sequential order of `IHostedService.StartAsync` calls so that hosted services are initialized / started in a pre-defined order? It looks like this should be the job of an `IHostBuilder` instance, but will there be a "standard" default implementation that preserves ordering?

This expands a general point that @jason-bragg was making above, drawing out specific experience from building the Orleans hosting mechanisms, but is also confirmed from experiences with several Java server systems I have built over the years.

**Comment by jthelin** [comment]
What is the purpose of the `CancellationToken` returned from `IApplicationLifetime.Application*` properties? Are they supposed to be like [old style] .NET Events that apps can register handlers for? 

Is `myApplicationLifetime.ApplicationStarted.IsCancellationRequested` condition becoming `true` similar to a [logical] `public event EventHandler ApplicationStarted` triggering? [modulo differences between push-vs-pull notification, of course.]

I think I am mostly getting confused by the choice of types used here because `IsCancellationRequested` seems like a "negative" assertion while `ApplicationStarted` seems like a "positive" assertion, although both allow similar `Register` operations for various "callback handlers" of course.

Also, does `CancellationToken.CanBeCancelled` have any meaning here distinct from `CancellationToken.IsCancellationRequested`?

**Comment by jthelin** [comment]
Is the `CancellationToken` param passed to the `IHostedService.[Start|Stop]Async` methods a complication that is useful or needed?

Say if a hosted service gets a call to `StopAsync` and starts to shut itself down, and then the cancellation token fires, is it supposed to "reinitialize" itself back to working state, ie. undo any [partial] shutdown changes it might be half way through?

If a `Stop` operation is allowed to be cancelled, then seems like for any non-trivial host system there are so many race conditions and corner-cases that would have to be considered and coded for that I am not sure even Einsteins would get it right.

It might be useful to think of a state transition diagram for lifecycle stages, to help clarify some of the edge-cases. For example, is this the expectation:
- StopAsync + Cancel => Running?
- StartAsync + Cancel => Stopped?

Is there a concrete IRL scenario for cancelling a server `Stop` operation in mid-flight, and what are the expected semantics that hosted services should aim for when that happens?

**Comment by davidfowl** [comment]
@jthelin 

> It is important for this spec to be clear about what ordering guarantees [or not] there are related to StartAsync / StopAsync call sequences. Specifically:

I'll update the spec with these details.

> Do calls to IHostedService.StartAsync happen sequentially, or in parallel? I am assuming sequential is the "default" [and much easier to reason about!], but is there also some way to have some services initialize in parallel with each other? [to expedite server startup times]

Sequentially. Parallel startup could be a bug farm.

> Do calls to IHostedService.StopAsync occur in reverse order to the StartAsync calls? Every system I have ever written what uses similar server-hosting functionality had some notion of "layering" which should be built up "bottom-to-top" and torn down "top-to-bottom".

We'll revisit this. Today I don't think we call stop in the reverse order.

> What is the mechanism to define & ensure a specific sequential order of IHostedService.StartAsync calls so that hosted services are initialized / started in a pre-defined order? It looks like this should be the job of an IHostBuilder instance, but will there be a "standard" default implementation that preserves ordering?

Today the DI container ensures order. Calls to register the IHostedService are ordered and will be preserved.

> What is the purpose of the CancellationToken returned from IApplicationLifetime.Application* properties? Are they supposed to be like [old style] .NET Events that apps can register handlers for?

Those are basically events. There were made CancellationTokens because it supports both callbacks and polling. Nothing is being cancelled, it's more of a better EventHandler.

> I think I am mostly getting confused by the choice of types used here because IsCancellationRequested seems like a "negative" assertion while ApplicationStarted seems like a "positive" assertion, although both allow similar Register operations for various "callback handlers" of course.

Yep, ignore that in this case. It could also be Task. See this https://github.com/dotnet/corefx/issues/16221.

> Also, does CancellationToken.CanBeCancelled have any meaning here distinct from CancellationToken.IsCancellationRequested?

Yes, they are different but they don't apply here.

> Is the CancellationToken param passed to the IHostedService.[Start|Stop]Async methods a complication that is useful or needed?

Yes, they are. CancellationTokens are cooperative, so if the caller tries to cancel start up or shutdown, the implementations should get a chance to react.

> Say if a hosted service gets a call to StopAsync and starts to shut itself down, and then the cancellation token fires, is it supposed to "reinitialize" itself back to working state, ie. undo any [partial] shutdown changes it might be half way through?

No, the CancellationToken is basically a timeout. The host says, I'll give you this long time shutdown and passes a token to StopAsync. If the token fires, it's the implementation's job to stop trying to shut down gracefully and just forcefully stop instead.

**Comment by jason-bragg** [comment]
@davidfowl, @andrebires @jthelin 
I agree with @jthelin that it is important that the spec/design be clear about startup/shutdown behaviors.

I am of the opinion that some sort of controlled (or staged) startup/shutdown is very useful, evidenced by both the need for this within Orleans, @andrebires system, and @jthelin's experiences.

As this sort of staged lifetime can be, IMO, introduced as a natural extension to the generic host, I'm not convinced it should be a first class requirement.  I'd rather see a solid minimal host implementation in place earlier than waiting for one with every bell and whistle (like staged lifetime) users may want.

I do, however, hope this capability is considered sufficiently in the initial version to allow its introduction by service developers or as a future improvement to the host.

**Comment by Antaris** [comment]
@davidfowl 

```
We only expose a Start and Stop event but other hosts might have more advanced events they want to surface. Windows store has suspend/resume, windows services can be paused and continued as well. Do we need to plan for expanding the life cycle or is that out of scope?
```

The HTTP context and server are both examples where you've introduced extensibility through the notion of features. If the new `IHost` was also extensible using the same/similar mechanism, other hosts/platforms could augment the host with support for things like Pause/Continue, etc.

That was you could introduce something like a `ISuspendable`, or something along those lines that allow you code to react if the host supports such scenarios?

**Comment by kijanawoodard** [comment]
This looks a lot like NServiceBus configuration. You might want to ping @andreasohlund and see what lessons they've learned. I think they're revamping their generic hosting approach.

**Comment by davidfowl** [comment]
> The HTTP context and server are both examples where you've introduced extensibility through the notion of features. If the new IHost was also extensible using the same/similar mechanism, other hosts/platforms could augment the host with support for things like Pause/Continue, etc.

Maybe, I'm not convinced yet though.

> That was you could introduce something like a ISuspendable, or something along those lines that allow you code to react if the host supports such scenarios?

How do you imagine that being used and what would the impact of such changes be on the interface?

> In terms of migrating IWebHost to IHost and IWebHostBuilder to IHostBuilder, could it not simply be:

The interfaces don't map very well (there are improvements we want to make) and this would be a breaking change.

**Comment by cwe1ss** [comment]
Thanks for pinging me David! Can't wait for this to be done - this will be extremely useful!!!

I don't understand the difference between `IApplicationLifetime.StopApplication` vs. `IHostLifetime.StopAsync`. Are both needed (isn't IApplicationLifetime a passive notification system?) and when would I use which?

There was a brief mention of the replacement for "Startup.Configure" (which would be called by a e.g. WebServiceHostedService), but I'm wondering what the replacement will be for the `IWebHostBuilder.Configure` method? Since this is a web/http-specific method, will this method already live in a "sub configuration" of the host builder? E.g.
```csharp
new HostBuilder()
  .UseHttpServer(options =>
  {
    options.Configure(app =>
    {
      app.UseMvc();
    }
  }
```

**Comment by cwe1ss** [comment]
My last question should probably be generalized to "Should all HTTP-specific extension methods that currently sit on IWebHostBuilder (e.g. Configure, UseStartup, ...) target IHostBuilder in the new system or should they be moved to a host-specific configuration (e.g. `UseHttpServer(options => options.Configure(...));`?

**Comment by guardrex** [comment]
> When `IHost.StopAsync` is called, the host will notify the `IHostLifetime` that it has stopped.

Seems strange to setup the `CancelKeyPress` event in an *On*Stopping method.

```csharp
public void OnStopping(Action<object> callback, object state)
{
    Console.CancelKeyPress += (sender, e) =>
    {
        e.Cancel = true;
        callback(state);
    };
}
```

**Comment by anurse** [comment]
A clearer name might be `AddOnStoppingHandler`. The method is run to register a callback to run when stopping, not when the host is actually stopping.

**Comment by anurse** [comment]
It's done that way so that the Host can register with the Lifetime so that it gets notified when the app is being stopped and shutdown all the services. For a console app, this is triggered by CancelKeyPress, for a Windows Service Lifetime, this would be registering a handler for when the user stops the service.

**Comment by Tratcher** [comment]
A branch is in progress here: https://github.com/aspnet/Hosting/tree/tratcher/generic. Everything is pretty much working. I agree that IHostLifetime is the most confusing element.

**Comment by Tratcher** [comment]
@cwe1ss I did one extension on IHostBuilder for adding an HTTP service that has a Configure lambda. Adding the server also moves to IHostBuilder because it only adds a service. As discussed above, UseStartup doesn't work well in this model due to the double service container requirement. What other HTTP specific extensions should we consider?

https://github.com/aspnet/Hosting/blob/79502594ce61bb5f56a4c7fa6f72cedf53bec97a/samples/GenericWebHost/WebHostExtensions.cs#L17

**Comment by guardrex** [comment]
1. Will u get to the point of having something like the convenience methods of `WebHost` (e.g., `CreateDefaultBuilder`)?
```csharp
public class Program
{
    public static async Task Main(string[] args)
    {
        await BuildHost(args).Run();
    }

    public static Task<IHost> BuildHost(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureAppConfiguration((hostContext, config) =>
            {
            })
            .ConfigureServices((hostContext, services) =>
            {
            })
            .ConfigureWebHost(app =>
            {
            })
            .Build();
}
```
2. If that's even remotely correct (:smile:), I wonder why "Configure*Web*Host" is there? ... I would have expected `.ConfigureHost()`.
2. Will the config chaining work out as I hope? Described in [Chained config (Configuration/630)](https://github.com/aspnet/Configuration/issues/630) (active issue) and [Can I get a value loaded by CreateDefaultBuilder for use in ConfigureAppConfiguration? (Configuration/707)](https://github.com/aspnet/Configuration/issues/707) (duplicate, but shows the pattern I hope for).

**Comment by anurse** [comment]
> I wonder why "ConfigureWebHost" is there? ... I would have expected .ConfigureHost()

Perhaps `ConfigureWebApplication` or `AddWebApplication` would be a better name. That method is adding the Web Hosting extensions to the generic host and configuring your application middleware/etc.

**Comment by guardrex** [comment]
> adding the Web Hosting extensions to the generic host and configuring your application middleware/etc.

Right, but they're not always for "web" hosting, right? E.g., some middleware thing that maybe just runs a bunch of background tasks. I asked in the sense of removing the word "web" from generic hosting system naming, so that's why I was a little surprised to see it change in the first place from just `Configure` to `ConfigureWebHost`, but in the second place not go from `Configure` to `ConfigureHost`. I just thought that the trend in the renaming of things for the new hosting was to get rid of the word "web."

**Comment by dls314** [comment]
> Right, but they're not always for "web" hosting, right? E.g., some middleware thing that maybe just runs a bunch of background tasks.

This is part, together with getting appropriate instances into all the right containers, that I'm interested in. The background task is something like the dispatching loop for a service bus. It would be nice to represent that background task using generic hosting.

**Comment by anurse** [comment]
> some middleware thing that maybe just runs a bunch of background tasks.

That's not a "Middleware" by our definition though. A Middleware is any function assignable to type [`RequestDelegate`](https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/RequestDelegate.cs). Middleware is not a generic term for extending an application, it's specific to fitting into some kind of **pipeline** (hence *Middle*ware). In order for Middleware to work, there needs to be someone pumping data through the pipeline (i.e. Kestrel).

> I just thought that the trend in the renaming of things for the new hosting was to get rid of the word "web."

Well, this isn't just a renaming "trend" ;). We're trying to make a very specific change here: We want to take the non-Web-specific parts of our existing hosting model (DI, Configuration, Options, Logging) and extract them into something generic. This generic thing could be used from basically ANY kind of .NET application (UWP, Command Line tools, etc.).

> in the second place not go from Configure to ConfigureHost

Something to make clear here: In the Generic Host, there is **no equivalent** to `Configure` by design. It simply doesn't exist. The `Configure` method, and the Middleware pipeline are specific to Web applications. We are doing some exploration into non-HTTP applications for Middleware (https://github.com/aspnet/KestrelHttpServer/issues/1980) but even then, it's not something that's generic to all applications, it's specific to a Server application of some kind. You wouldn't register Middleware into a UWP app or a Command Line tool, etc.

So, the `Configure` method only appears when you attach the Web Hosting components to your Generic Host-based application, hence putting it in `.ConfigureWebHost` (and one reason I kinda prefer a name like `.AddWebHosting` because you're actually adding the Web Hosting components to the container in that method, not just configuring them)

**Comment by guardrex** [comment]
> That's not a "Middleware" by our definition though.

I meant (and should've typed out): "that runs in response to a server-processed non-HTTP requests." I envision some kind of task processor (run by middleware) that could process requests and fire off tasks ... not web/HTTP requests and not returning web/HTTP responses.

I see what u mean ...

> ... there needs to be someone pumping data through the pipeline ...

>  In the Generic Host, there is no equivalent to `Configure` by design. 

> ... the Middleware pipeline are specific to Web applications.

> ... the Configure method only appears when you attach the Web Hosting components to your Generic Host-based application ...

All good. :+1: I just hope ur correct about the stability of this. If there's any chance at all that server+pipeline processing would be non-"web" (non-HTTP) in the future, then it *seems like* pinning to the word "web" there might result in a name change later. I guess it's all well designed and won't change. Thanks for explaining the reasoning out.

**Comment by anurse** [comment]
> If there's any chance at all that server+pipeline processing would be non-"web" (non-HTTP) in the future, then it seems like pinning to the word "web" there might result in a name change later.

This is totally valid. We are in fact looking at non-HTTP pipelines and what those look like. Having said that, we'll still want to preserve the simplicity of our current set-up for launching ASP.NET apps, so we'll always want a `.ConfigureWebHost` extension method of some kind. It may be that later on there's an `.ConfigureServerHost` that's more generic (and `.ConfigureWebHost(...)` becomes a synonym for `.ConfigureServerHost(serverBuilder => serverBuilder.UseHttp(...))` or something), but I don't think there's a risk of burning anything we haven't already burned. `IApplicationBuilder` will always be associated with HTTP applications because of how massive a breaking change it would be to untie that, so associating it with the "Web" term won't limit it any more than it already is.

**Comment by anurse** [comment]
> I envision some kind of task processor (run by middleware) that could process requests and fire off tasks

Yeah, we've talked about this, but the current Middleware pipeline is so coupled to a client-driven request/response model that it doesn't generalize well into other protocols. Many non-HTTP protocols allow the server to send unsolicited data to the client, or send multiple responses to a single message, or not send a response at all, which doesn't map into our current Middleware pipeline at all. The current pipeline is very much driven by the HTTP paradigm. So anything generic we do is going to have to sit at a different level than our HTTP pipeline anyway.

[Project Bedrock](https://github.com/aspnet/KestrelHttpServer/issues/1980) is focusing on the non-HTTP stuff and we've focused on Connection-level Middleware that operates on the raw stream as a whole (i.e. SSL/TLS, Connection Limits, Flow Control, etc.) right now because of how difficult it is to build a generic "message processing" pipeline.

**Comment by muratg** [comment]
cc @ajcvickers who brought up the point that we'll need to think through design-time issues as well. (Like EF migrations.)

**Comment by muratg** [comment]
Removing `needs design` label. This work is undergoing.

**Comment by Tratcher** [comment]
Please take a look at the PR for this feature: https://github.com/aspnet/Hosting/pull/1204

**Comment by davidfowl** [comment]
We have an implementation now. Thanks for the feedback! Please try it out and file issues.

**Comment by sirajmansour** [comment]
> A note about Startup classes - While the concept of a Startup class is extremely useful, it's also very problematic as it requires
2 dependency injection containers to be built during the startup cycle. This causes issues where singletons aren't the same when activating the
Startup class and when activating an IHostedService. This is because Startup.ConfigureServices lets the user add more services but also lets them
get at hosting services in the Startup.ctor.

@davidfowl i am assuming this is why a singleton injected into my `IHostedService` is a different instance than the one resolve and configure somewhere else. Is there a work around that you can suggest ?

**Comment by muratg** [comment]
@Tratcher / @davidfowl -- was it just an issue of seperation of concerns?

**Comment by poke** [comment]
My guess is that this is due to how the configuration evolved over time. Back in the days™ we had a `hosting.json` by default that was responsible of setting up hosting related configuration, and the configuration was completely separated with the application specific stuff. It only happened later that the `hosting.json` was removed, and you were supposed to do that yourself. And only with the new web host builder, hosting related stuff like logging or the 2.1 Kestrel configuration also moved into the `appsettings.json`, making a single configuration builder more appropriate.

**Comment by Tratcher** [comment]
Some minimal amount of config is needed to bootstrap the host, IHostingEnvironment, etc.. That information is then used as input to the rest of the HostBuilder flow. E.g. `configApp.AddJsonFile(
                $"appsettings.{hostContext.HostingEnvironment.EnvironmentName}.json",` wouldn't work with a single config stage, you don't know where to get EnvironmentName from.

**Comment by Tratcher** [comment]
@guardrex note we made some updates to IHostLifetime after we had trouble getting it to work with windows services. See if the new version makes any more sense: https://github.com/aspnet/Hosting/blob/38f691c09e3aae4f34be139c39ae1fb264803fd0/src/Microsoft.Extensions.Hosting/Internal/ConsoleLifetime.cs

**Comment by jwfx** [comment]
@Tratcher 
I'm just facing the problem that IApplicationLifetime.StopApplication() does not stop an application that is started as a Windows service (`public class WindowsService : WebHostService`). Currently using 2.0 bits. 

Are these updates related to this issue and if yes, I suppose they are only available in 2.1?

**Comment by Tratcher** [comment]
@jwfx that's not related to this issue, please open a new one and give the details for your scenario.

**Comment by kosa-gyula-77** [comment]
Re the issues where singletons aren't the same when activating the Startup class and when activating an IHostedService, are web host builders supposed to 1) switch to building their host with the services of generic host or 2) just upgrade their use of aspnet/Hosting to version 2.1.0+ , please?
